File: mac/AirBridgeMac/Data/Relay/RelayHTTPClient.swift
import Foundation

protocol RelayPairingClient {
    func createPairingSessionResponse(
        deviceName: String,
        platform: AirBridgePlatform,
        publicKey: Data
    ) async throws -> CreatePairingSessionResponse

    func lookupPairingSession(
        sessionID: String,
        pairingSecret: String
    ) async throws -> PairingSessionSnapshot

    func completePairingSession(
        sessionID: String,
        pairingSecret: String
    ) async throws -> CompletePairingSessionResponse
}

enum RelayHTTPClientError: LocalizedError {
    case invalidBaseURL
    case invalidResponse
    case serviceError(RelayServiceErrorPayload)

    var errorDescription: String? {
        switch self {
        case .invalidBaseURL:
            return "Relay URL이 올바르지 않아요."
        case .invalidResponse:
            return "Relay가 올바르지 않은 HTTP 응답을 반환했어요."
        case .serviceError(let payload):
            return payload.message
        }
    }
}

final class RelayHTTPClient: RelayPairingClient {
    private let baseURL: URL
    private let session: URLSession

    init(baseURL: URL, session: URLSession = .shared) {
        self.baseURL = baseURL
        self.session = session
    }

    func createPairingSessionResponse(
        deviceName: String,
        platform: AirBridgePlatform,
        publicKey: Data
    ) async throws -> CreatePairingSessionResponse {
        let validatedDeviceName = try RelayInputValidator.deviceName(deviceName)
        let validatedPublicKey = try RelayInputValidator.publicKey(publicKey)

        let response: CreatePairingSessionResponse = try await send(
            path: "/api/v1/pairing/sessions",
            method: "POST",
            body: CreatePairingSessionRequest(
                deviceName: validatedDeviceName,
                platform: platform.rawValue,
                publicKey: validatedPublicKey.rawBase64EncodedString
            )
        )

        return response
    }

    func lookupPairingSession(
        sessionID: String,
        pairingSecret: String
    ) async throws -> PairingSessionSnapshot {
        let validatedSessionID = try RelayInputValidator.identifier(sessionID, field: "pairing_session_id")
        let validatedPairingSecret = try RelayInputValidator.pairingSecret(pairingSecret)

        let response: LookupPairingSessionResponse = try await send(
            path: "/api/v1/pairing/sessions/\(validatedSessionID)/lookup",
            method: "POST",
            body: LookupPairingSessionRequest(pairingSecret: validatedPairingSecret)
        )

        return PairingSessionSnapshot(
            pairingSessionID: response.pairingSessionID,
            state: response.state,
            initiatorDeviceID: response.initiatorDeviceID,
            initiatorName: response.initiatorName,
            initiatorPlatform: response.initiatorPlatform,
            initiatorPublicKey: try Data(rawBase64Encoded: response.initiatorPublicKey),
            joinerDeviceID: response.joinerDeviceID,
            joinerName: response.joinerName,
            joinerPlatform: response.joinerPlatform,
            joinerPublicKey: try response.joinerPublicKey.map(Data.init(rawBase64Encoded:)),
            expiresAt: response.expiresAt,
            updatedAt: response.updatedAt,
            completedAt: response.completedAt
        )
    }

    func completePairingSession(
        sessionID: String,
        pairingSecret: String
    ) async throws -> CompletePairingSessionResponse {
        let validatedSessionID = try RelayInputValidator.identifier(sessionID, field: "pairing_session_id")
        let validatedPairingSecret = try RelayInputValidator.pairingSecret(pairingSecret)

        let response: CompletePairingSessionResponse = try await send(
            path: "/api/v1/pairing/sessions/\(validatedSessionID)/complete",
            method: "POST",
            body: CompletePairingSessionRequest(pairingSecret: validatedPairingSecret)
        )

        return response
    }

    private func send<Body: Encodable, Response: Decodable>(
        path: String,
        method: String,
        body: Body
    ) async throws -> Response {
        let requestURL = try makeURL(path: path)
        var request = URLRequest(url: requestURL)
        request.httpMethod = method
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue("application/json", forHTTPHeaderField: "Accept")
        request.httpBody = try JSONEncoder.airBridge.encode(body)

        let (data, response) = try await session.data(for: request)
        return try decodeResponse(data: data, response: response)
    }

    private func decodeResponse<Response: Decodable>(
        data: Data,
        response: URLResponse
    ) throws -> Response {
        guard let httpResponse = response as? HTTPURLResponse else {
            throw RelayHTTPClientError.invalidResponse
        }

        guard (200 ... 299).contains(httpResponse.statusCode) else {
            if let errorResponse = try? JSONDecoder.airBridge.decode(RelayHTTPErrorResponse.self, from: data) {
                throw RelayHTTPClientError.serviceError(errorResponse.error)
            }

            throw RelayHTTPClientError.invalidResponse
        }

        return try JSONDecoder.airBridge.decode(Response.self, from: data)
    }

    private func makeURL(path: String) throws -> URL {
        guard var components = URLComponents(url: baseURL, resolvingAgainstBaseURL: false) else {
            throw RelayHTTPClientError.invalidBaseURL
        }

        let normalizedBasePath = components.path.hasSuffix("/") ? String(components.path.dropLast()) : components.path
        components.path = normalizedBasePath + path

        guard let url = components.url else {
            throw RelayHTTPClientError.invalidBaseURL
        }

        return url
    }
}

private struct CreatePairingSessionRequest: Encodable {
    let deviceName: String
    let platform: String
    let publicKey: String

    enum CodingKeys: String, CodingKey {
        case deviceName = "device_name"
        case platform
        case publicKey = "public_key"
    }
}

struct CreatePairingSessionResponse: Decodable {
    let pairingSessionID: String
    let pairingSecret: String
    let initiatorDeviceID: String
    let initiatorRelayToken: String
    let expiresAt: Date

    enum CodingKeys: String, CodingKey {
        case pairingSessionID = "pairing_session_id"
        case pairingSecret = "pairing_secret"
        case initiatorDeviceID = "initiator_device_id"
        case initiatorRelayToken = "initiator_relay_token"
        case expiresAt = "expires_at"
    }
}

private struct LookupPairingSessionRequest: Encodable {
    let pairingSecret: String

    enum CodingKeys: String, CodingKey {
        case pairingSecret = "pairing_secret"
    }
}

private struct LookupPairingSessionResponse: Decodable {
    let pairingSessionID: String
    let state: PairingSessionState
    let initiatorDeviceID: String
    let initiatorName: String
    let initiatorPlatform: AirBridgePlatform
    let initiatorPublicKey: String
    let joinerDeviceID: String?
    let joinerName: String?
    let joinerPlatform: AirBridgePlatform?
    let joinerPublicKey: String?
    let expiresAt: Date
    let updatedAt: Date
    let completedAt: Date?

    enum CodingKeys: String, CodingKey {
        case pairingSessionID = "pairing_session_id"
        case state
        case initiatorDeviceID = "initiator_device_id"
        case initiatorName = "initiator_name"
        case initiatorPlatform = "initiator_platform"
        case initiatorPublicKey = "initiator_public_key"
        case joinerDeviceID = "joiner_device_id"
        case joinerName = "joiner_name"
        case joinerPlatform = "joiner_platform"
        case joinerPublicKey = "joiner_public_key"
        case expiresAt = "expires_at"
        case updatedAt = "updated_at"
        case completedAt = "completed_at"
    }
}

struct CompletePairingSessionResponse: Decodable {
    let pairingSessionID: String
    let state: PairingSessionState
    let completedAt: Date

    enum CodingKeys: String, CodingKey {
        case pairingSessionID = "pairing_session_id"
        case state
        case completedAt = "completed_at"
    }
}

private struct CompletePairingSessionRequest: Encodable {
    let pairingSecret: String

    enum CodingKeys: String, CodingKey {
        case pairingSecret = "pairing_secret"
    }
}

File: android/app/src/main/java/com/airbridge/app/data/relay/RelayHttpClient.kt
package com.airbridge.app.data.relay

import com.airbridge.app.domain.PairingJoinResult
import com.airbridge.app.domain.PairingSessionSnapshot
import com.airbridge.app.domain.PairingSessionState
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody

class RelayHttpClient(
    private val okHttpClient: OkHttpClient,
    private val json: Json,
) {
    suspend fun findWorkingRelayUrl(
        relayAddresses: List<String>,
    ): String = withContext(Dispatchers.IO) {
        require(relayAddresses.isNotEmpty()) { "relay 주소 목록이 비어 있어요" }

        for (address in relayAddresses) {
            val success = tryConnectRelay(address)
            if (success) {
                return@withContext address
            }
        }

        throw RelayApiException(
            code = "connection_failed",
            message = "어떤 relay 주소에도 연결할 수 없어요. 주소 목록: $relayAddresses",
            statusCode = 0,
        )
    }

    private suspend fun tryConnectRelay(baseUrl: String): Boolean = withContext(Dispatchers.IO) {
        runCatching {
            val request = Request.Builder()
                .url("${baseUrl.normalizedBaseUrl()}/api/v1/server/info")
                .get()
                .build()
            okHttpClient.newCall(request).execute().use { response ->
                response.isSuccessful
            }
        }.getOrDefault(false)
    }

    suspend fun joinPairingSession(
        relayBaseUrl: String,
        pairingSessionId: String,
        pairingSecret: String,
        deviceName: String,
        publicKeyBase64: String,
    ): PairingJoinResult = withContext(Dispatchers.IO) {
        require(deviceName.length <= RelayServerLimits.MAX_DEVICE_NAME_LENGTH) {
            "기기 이름은 ${RelayServerLimits.MAX_DEVICE_NAME_LENGTH}자를 넘을 수 없어요 (현재: ${deviceName.length}자)"
        }
        require(pairingSecret.length <= RelayServerLimits.MAX_PAIRING_SECRET_LENGTH) {
            "페어링 비밀값은 ${RelayServerLimits.MAX_PAIRING_SECRET_LENGTH}자를 넘을 수 없어요 (현재: ${pairingSecret.length}자)"
        }
        
        val request = JoinPairingSessionRequest(
            pairingSecret = pairingSecret,
            deviceName = deviceName,
            platform = "android",
            publicKey = publicKeyBase64,
        )
        val response = executeJson<JoinPairingSessionResponse>(
            url = "${relayBaseUrl.normalizedBaseUrl()}/api/v1/pairing/sessions/$pairingSessionId/join",
            method = "POST",
            body = request,
            serializer = JoinPairingSessionResponse.serializer(),
        )
        response.toDomain()
    }

    suspend fun lookupPairingSession(
        relayBaseUrl: String,
        pairingSessionId: String,
        pairingSecret: String,
    ): PairingSessionSnapshot = withContext(Dispatchers.IO) {
        require(pairingSecret.length <= RelayServerLimits.MAX_PAIRING_SECRET_LENGTH) {
            "페어링 비밀값은 ${RelayServerLimits.MAX_PAIRING_SECRET_LENGTH}자를 넘을 수 없어요 (현재: ${pairingSecret.length}자)"
        }
        
        val response = executeJson<LookupPairingSessionResponse>(
            url = "${relayBaseUrl.normalizedBaseUrl()}/api/v1/pairing/sessions/$pairingSessionId/lookup",
            method = "POST",
            body = LookupPairingSessionRequest(pairingSecret = pairingSecret),
            serializer = LookupPairingSessionResponse.serializer(),
        )
        response.toDomain()
    }

    suspend fun completePairingSession(
        relayBaseUrl: String,
        pairingSessionId: String,
        pairingSecret: String,
    ): CompletePairingSessionResponse = withContext(Dispatchers.IO) {
        require(pairingSecret.length <= RelayServerLimits.MAX_PAIRING_SECRET_LENGTH) {
            "페어링 비밀값은 ${RelayServerLimits.MAX_PAIRING_SECRET_LENGTH}자를 넘을 수 없어요 (현재: ${pairingSecret.length}자)"
        }
        
        executeJson(
            url = "${relayBaseUrl.normalizedBaseUrl()}/api/v1/pairing/sessions/$pairingSessionId/complete",
            method = "POST",
            body = CompletePairingSessionRequest(pairingSecret = pairingSecret),
            serializer = CompletePairingSessionResponse.serializer(),
        )
    }

    private suspend fun <T> executeJson(
        url: String,
        method: String,
        body: Any?,
        serializer: kotlinx.serialization.KSerializer<T>,
    ): T = withContext(Dispatchers.IO) {
        val requestBuilder = Request.Builder().url(url)
        val requestBody = body?.let {
            json.encodeToString(
                serializer = serializerForBody(it),
                value = it,
            ).toRequestBody(JsonMediaType)
        }

        when (method) {
            "POST" -> requestBuilder.post(requireNotNull(requestBody))
            "GET" -> requestBuilder.get()
            else -> error("지원하지 않는 HTTP 메서드예요: $method")
        }

        okHttpClient.newCall(requestBuilder.build()).execute().use { response ->
            val rawBody = response.body?.string().orEmpty()
            if (!response.isSuccessful) {
                val errorResponse = runCatching {
                    json.decodeFromString(RelayErrorEnvelope.serializer(), rawBody)
                }.getOrNull()
                throw RelayApiException(
                    code = errorResponse?.error?.code ?: "http_${response.code}",
                    message = errorResponse?.error?.message ?: "relay 서버 요청에 실패했어요",
                    statusCode = response.code,
                )
            }

            return@use json.decodeFromString(serializer, rawBody)
        }
    }

    @Suppress("UNCHECKED_CAST")
    private fun <T : Any> serializerForBody(value: T): kotlinx.serialization.KSerializer<T> {
        return when (value) {
            is JoinPairingSessionRequest -> JoinPairingSessionRequest.serializer() as kotlinx.serialization.KSerializer<T>
            is LookupPairingSessionRequest -> LookupPairingSessionRequest.serializer() as kotlinx.serialization.KSerializer<T>
            is CompletePairingSessionRequest -> CompletePairingSessionRequest.serializer() as kotlinx.serialization.KSerializer<T>
            else -> error("지원하지 않는 요청 본문 타입이에요: ${value::class.java.name}")
        }
    }

    companion object {
        private val JsonMediaType = "application/json; charset=utf-8".toMediaType()
    }
}

class RelayApiException(
    val code: String,
    override val message: String,
    val statusCode: Int,
) : Exception(message)

private fun String.normalizedBaseUrl(): String = trim().removeSuffix("/")

@Serializable
private data class JoinPairingSessionRequest(
    @SerialName("pairing_secret")
    val pairingSecret: String,
    @SerialName("device_name")
    val deviceName: String,
    val platform: String,
    @SerialName("public_key")
    val publicKey: String,
)

@Serializable
private data class JoinPairingSessionResponse(
    @SerialName("pairing_session_id")
    val pairingSessionId: String,
    @SerialName("joiner_device_id")
    val joinerDeviceId: String,
    @SerialName("joiner_relay_token")
    val joinerRelayToken: String,
    @SerialName("initiator_device_id")
    val initiatorDeviceId: String,
    @SerialName("initiator_public_key")
    val initiatorPublicKey: String,
    @SerialName("expires_at")
    val expiresAt: String,
) {
    fun toDomain(): PairingJoinResult = PairingJoinResult(
        pairingSessionId = pairingSessionId,
        joinerDeviceId = joinerDeviceId,
        joinerRelayToken = joinerRelayToken,
        initiatorDeviceId = initiatorDeviceId,
        initiatorPublicKey = initiatorPublicKey,
        expiresAt = expiresAt,
    )
}

@Serializable
private data class LookupPairingSessionRequest(
    @SerialName("pairing_secret")
    val pairingSecret: String,
)

@Serializable
private data class LookupPairingSessionResponse(
    @SerialName("pairing_session_id")
    val pairingSessionId: String,
    val state: String,
    @SerialName("initiator_device_id")
    val initiatorDeviceId: String,
    @SerialName("initiator_name")
    val initiatorName: String,
    @SerialName("initiator_platform")
    val initiatorPlatform: String,
    @SerialName("initiator_public_key")
    val initiatorPublicKey: String,
    @SerialName("joiner_device_id")
    val joinerDeviceId: String? = null,
    @SerialName("joiner_name")
    val joinerName: String? = null,
    @SerialName("joiner_platform")
    val joinerPlatform: String? = null,
    @SerialName("joiner_public_key")
    val joinerPublicKey: String? = null,
    @SerialName("expires_at")
    val expiresAt: String,
    @SerialName("updated_at")
    val updatedAt: String,
    @SerialName("completed_at")
    val completedAt: String? = null,
) {
    fun toDomain(): PairingSessionSnapshot = PairingSessionSnapshot(
        pairingSessionId = pairingSessionId,
        state = PairingSessionState.fromWire(state),
        initiatorDeviceId = initiatorDeviceId,
        initiatorName = initiatorName,
        initiatorPlatform = initiatorPlatform,
        initiatorPublicKey = initiatorPublicKey,
        joinerDeviceId = joinerDeviceId,
        joinerName = joinerName,
        joinerPlatform = joinerPlatform,
        joinerPublicKey = joinerPublicKey,
        expiresAt = expiresAt,
        updatedAt = updatedAt,
        completedAt = completedAt,
    )
}

@Serializable
private data class CompletePairingSessionRequest(
    @SerialName("pairing_secret")
    val pairingSecret: String,
)

@Serializable
data class CompletePairingSessionResponse(
    @SerialName("pairing_session_id")
    val pairingSessionId: String,
    val state: String,
    @SerialName("completed_at")
    val completedAt: String,
)

@Serializable
private data class RelayErrorEnvelope(
    val error: RelayErrorBody,
)

@Serializable
private data class RelayErrorBody(
    val code: String,
    val message: String,
)

